### PR TITLE
Dev/yingqi

### PR DESCRIPTION
**Pull request by ioeddk**
Added pre-training scripts, fine-tuning scripts, evaluation scripts, and some bash scripts to the pipeline of fine-tuning downstream tasks.

### CODE REVIEW

File: hmt_src/main.py [hunk 1]
```diff
@@ -93,7 +121,8 @@ def main():
     global torch
 
     args = parser.parse_args()
-    accelerator = Accelerator()
+    ddp_kwargs = DistributedDataParallelKwargs(find_unused_parameters=True)
```
Comment (OswaldHe) [inline hmt_src/main.py:125]: Will that prevent people from using ZeRO3 or FSDP?

File: hmt_src/main.py [hunk 2]
```diff
@@ -82,7 +108,9 @@
 parser.add_argument('--train_memory_map', action='store_true', default=False, help='train memory projection for dynamic reading speed.')
 parser.add_argument('--inject_autoencoder', action='store_true', default=False, help='use autoencoder to compress/decompress the intermediate embeddings.')
 parser.add_argument('--generate', type=str, default=None, help='generate for harry potter book.')
-
+parser.add_argument('--streaming', action='store_true', default=False, help='generate text in streaming mode')
+parser.add_argument('--shuffle', action='store_true', default=False, help='shuffle the dataset')
```
Comment (OswaldHe) [inline hmt_src/main.py:112]: The default should be true for the training set and false for valid and test tests.

File: scripts/eval/generate/generate_qwen2.5-0.5b.sh [hunk 3]
```diff
@@ -0,0 +1,38 @@
+#!/bin/bash
+
+
+
+export NCCL_DEBUG=INFO
+export TORCH_DISTRIBUTED_DEBUG=INFO
+
+accelerate envexport WEIGHT_BASE=hmt_pretrained/qwen2.5-0.5b
```
Comment (OswaldHe) [inline scripts/eval/generate/generate_qwen2.5-0.5b.sh]: Use shell script positional argument. Default as hmt_pretrained/qwen2.5-0.5b

File: scripts/eval/generate/generate_qwen2.5-0.5b.sh [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+            --task_subset=sample \
+            --training_step=100 \
+            --num_sensory=32 \
+            --segment_length=512 \
+            --bptt_depth=6 \
+            --train_set_split=2 \
+            --num_seg_save=8 \
+            --batch_size=2 \
+            --test_length=10000 \
+            --save_dir=checkpoints/rp_opt-350m \
```
Comment (OswaldHe) [inline scripts/eval/generate/generate_qwen2.5-0.5b.sh:46]: in generate.py, check if directory exists and create one if not.

File: tools/data_processing/cache.py
Comment (OswaldHe) [inline tools/data_processing/cache.py:1]: delete empty file

File: tools/data_processing/prep_funcs.py [hunk 5]
```diff
@@ -0,0 +1,46 @@
+import datasets
+
+def prepare_narrativeqa(dataset: datasets.Dataset):
```
Comment (OswaldHe) [inline tools/data_processing/prep_funcs.py:3]: move the prepare function in musique and qmsum here

**Review by OswaldHe — CHANGES_REQUESTED** [review]
Please propagate changes to other files

**Comment by OswaldHe** [comment]
Also, add documentation about executing training/finetuning/evaluation/generation scripts.

File: hmt_src/main.py (hunk 2 above)
Comment (ioeddk) [reply hmt_src/main.py:112]: add a shuffle option for the training set.

File: scripts/eval/generate/generate_qwen2.5-0.5b.sh (hunk 3 above)
Comment (ioeddk) [reply scripts/eval/generate/generate_qwen2.5-0.5b.sh]: OK, only for checkpoint. Also, pass in save_dir, check existence, and create if not exist. Use current one as default.

File: tools/data_processing/prep_funcs.py (hunk 5 above)
Comment (ioeddk) [reply tools/data_processing/prep_funcs.py:3]: Sure.

**Review by OswaldHe — APPROVED** [review]
lgtm
